Add support to take key files with windows new line ending

bingbing8 opened the pull request:
1. Add support to take key files with windows new line ending (https://github.com/PowerShell/Win32-OpenSSH/issues/1130)
2. add test cases for CRLF
3. Update test helper script to catch the exitcode of unittest and report the failure
4. Enable uni test unittest-sshkey and unittest-sshkey
5. Disable resource check for signal tests due to some API issue to follow.
6. Remove workaround for windows new line ending in test scripts
7. Add test validation for ACL of registry entries when perform ssh-add

bingbing8 committed:
update test helper script to catch the exitcode of unittest and report the failure

bingbing8 committed:
add RDP to investigate unit test failure

bingbing8 committed:
test if only the first time failed

bingbing8 committed:
update for troubleshooting

bingbing8 committed:
comment out some TEST_RESOURCES calls

NoMoreFood commented on regress/unittests/win32compat/signal_tests.c:
```diff
@@ -160,7 +160,7 @@ signal_test_wait_for_multiple_objects()
 		}
 
-		TEST_RESOURCES(FALSE);
+		//TEST_RESOURCES(FALSE);
```
@bingbing8 These were my invention to detect resource leaks.  Were these tests failing for you?  If so, I have some investigation to do.  I ran the unit tests under both release/debug mode without any issue on my side.

bingbing8 replied:
@NoMoreFood, yes, tests failed on win2k12r2 especially 100% failed on CI system if I enabled these lines. please see [test run for previous commit in this PR](https://ci.appveyor.com/project/PowerShell/openssh-portable/build/7.7.0.0.1307). It randomly failed on my local win2k12r2 machine and does not repro on win10. unit test failures were not reported correctly because exit code is not caught previously, but you can see unit test unittest-win32compat stop reporting as part of unittestresults.txt starting [PR 284]( https://github.com/PowerShell/openssh-portable/pull/284). unit test output can be found at [here](https://ci.appveyor.com/project/PowerShell/openssh-portable/build/7.6.0.0.1231/artifacts)
I disabled them just for troubleshooting purpose and also take advantage of the CI test run. If you can investigate the issue, that would be great. You can enable these line to repro the issue. to Access the CI system machine, you can add [RDP access](https://github.com/PowerShell/openssh-portable/pull/301/commits/b6d345879ab79c9f6f308f2e279027b6100da690). Thanks!

NoMoreFood replied:
@bingbing8 Thanks, I'll investigate.

NoMoreFood replied:
@bingbing8 This looks like a consistency issue with the value that GetProcessHandleCount() returns, specifically on Windows 8.1 / 2012 R2.  So for now it seems to be fine to keep it commented out.   Do you happen to know anyone in the kernel development area?  If so, could you see if the attached code makes any sense to them?  On Windows 10, before and after handle count are identical.   On my Windows 8.1, the after count is one higher than the before count.  

[TestCode.txt](https://github.com/PowerShell/openssh-portable/files/1894624/TestCode.txt)

Update: Did some additional debugging.  Per below, it appears something internally is leaking event handles.  Depending on the run, the test code might leak 0, 1, or 2 event handles.  However, running the main() inner code 1000 times in a loop doesn't leak more 2 than event handles in total.  Seems like something internal to CreateThread is using events and cleaning them up in an inconsistent manner -- very unusual for unmanaged C code.

```
  Handle  Address Object Type --- Before
       4 FFFFE000CE2FF820 File (30)
       8 FFFFC00029298EA0 Directory (3)
       C FFFFE000CE632070 File (30)
      10 FFFFE000D1DBFD00 File (30)
      14 FFFFC00036412240 Key (37)
      18 FFFFE000D19AE960 File (30)
      1C FFFFE000CE427F20 File (30)
      20 FFFFE000CE427F20 File (30)
      24 FFFFE000CEB698C0 ALPC Port (38)
      28 FFFFC0002C06BA00 Key (37)
  Handle  Address Object Type --- After
       4 FFFFE000CE2FF820 File (30)
       8 FFFFC00029298EA0 Directory (3)
       C FFFFE000CE632070 File (30)
      10 FFFFE000D1DBFD00 File (30)
      14 FFFFC00036412240 Key (37)
      18 FFFFE000D19AE960 File (30)
      1C FFFFE000CE427F20 File (30)
      20 FFFFE000CE427F20 File (30)
      24 FFFFE000CEB698C0 ALPC Port (38)
      28 FFFFC0002C06BA00 Key (37)
      34 FFFFE000CEFEACA0 Event (12)
      38 FFFFE000D1C141B0 Event (12)
```

bingbing8 replied:
@NoMoreFood , I am following up the potential leaking issue with kernel development team. I will comment out these validation for now.

bingbing8 committed:
Enable tests unittest-sshkey and unittest-kex

bingbing8 committed:
1. add support for windows new line ending
2. remove workaround for windows new line ending in test scripts
3. add test cases for windows newline ending

bingbing8 committed:
minor updates

bingbing8 committed:
fix indent

bingbing8 committed:
remove the condition because the bug has fixed in product code

bingbing8 committed:
added CRLF definition

bingbing8 committed:
remove the init stage

manojampalam commented on contrib/win32/openssh/config.h.vs:
```diff
@@ -1580,6 +1580,7 @@
 /* #undef socklen_t */
 #define WIN32_LEAN_AND_MEAN 1
 #define WINDOWS 1
+#define CRLF 1
```
call it SUPPORT_CRLF

bingbing8 replied:
fixed

manojampalam commented on sshkey.c:
```diff
@@ -65,6 +65,12 @@
 #define MARK_END		"-----END OPENSSH PRIVATE KEY-----\n"
 #define MARK_BEGIN_LEN		(sizeof(MARK_BEGIN) - 1)
 #define MARK_END_LEN		(sizeof(MARK_END) - 1)
+#ifdef WINDOWS
```
use the CRLF macro

bingbing8 replied:
fixed

manojampalam commented on sshkey.c:
```diff
@@ -65,6 +65,12 @@
+#define MARK_END_CRLF		"-----END OPENSSH PRIVATE KEY-----\r\n"
+#define MARK_BEGIN_LEN_CRLF		(sizeof(MARK_BEGIN_CRLF) - 1)
+#define MARK_END_LEN_CRLF		(sizeof(MARK_END_CRLF) - 1)
+#endif // WINDOWS
```
replace //windows comment with crlf

bingbing8 replied:
fixed

bingbing8 committed:
update macro name SUPPORT_CRLF

manojampalam commented on sshkey.c:
```diff
@@ -3415,8 +3421,16 @@ sshkey_parse_private2(struct sshbuf *blob, int type, const char *passphrase,
+	    (encoded_len < (MARK_BEGIN_LEN_CRLF + MARK_END_LEN_CRLF) ||
+	    memcmp(cp, MARK_BEGIN_CRLF, MARK_BEGIN_LEN_CRLF) != 0)) {
+#else  /* !SUPPORT_CRLF */
 	if (encoded_len < (MARK_BEGIN_LEN + MARK_END_LEN) ||
```
Missed this in earlier review. So you want to support only one of the format at any time? Can we support both formats on Windows?

bingbing8 replied:
this change support both formats on windows

bingbing8 committed:
add validation  ACL for registry entries
